File: packages/node/src/sdk/OtlpTransport.ts
import { request } from 'node:http';
import { request as requestSecurely } from 'node:https';

import { Headers, LogEvent } from '../api/types';
import { Dt3ErrorCode, Dt3TransportError } from './errors';

/**
 * Raised when an OTLP export cannot be initiated successfully.
 */
export class OtlpTransportError extends Dt3TransportError {
  /**
   * Create an OTLP export error with a sanitized failure description.
   *
   * @param message - Safe transport failure description that never includes event data.
   */
  constructor(
    message: string,
    options: {
      code?: Dt3ErrorCode;
      retryable?: boolean;
      cause?: unknown;
    } = {},
  ) {
    super(message, options);
    this.name = 'OtlpTransportError';
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

/**
 * Synchronously initiates OTLP/HTTP JSON export for final DT3 structured log events.
 *
 * Node's HTTP client sends requests asynchronously. This transport does not buffer
 * events, and logger-level `fail_open` controls handling of synchronous initiation
 * failures without recursively logging transport errors.
 */
export class OtlpTransport {
  private static readonly severityNumbers: Record<string, number> = {
    TRACE: 1,
    DEBUG: 5,
    INFO: 9,
    WARN: 13,
    WARNING: 13,
    ERROR: 17,
    FATAL: 21,
  };

  private readonly endpoint: URL;
  private readonly timeoutMs: number;
  private readonly headers: Headers;
  private readonly inFlight = new Set<Promise<void>>();
  private readonly failedDeliveries = new Map<Promise<void>, unknown>();
  private readonly onDeliveryFailure?: (error: unknown) => void;
  private readonly retainFailedDeliveries: boolean;
  private closed = false;

  /**
   * Create an OTLP/HTTP JSON transport.
   *
   * @param endpoint - OTLP Logs endpoint, conventionally ending in `/v1/logs`.
   * @param timeoutMs - Maximum request duration in milliseconds.
   * @param headers - Optional request headers merged with the OTLP JSON content type.
   * @param onDeliveryFailure - Optional out-of-band async failure observer.
   * @param retainFailedDeliveries - Whether failures must be retained for a
   * later fail-closed `flush()` boundary.
   * @throws Error if endpoint, timeout, or headers are invalid.
   */
  constructor(
    endpoint: string,
    timeoutMs = 10000,
    headers?: Headers,
    onDeliveryFailure?: (error: unknown) => void,
    retainFailedDeliveries = true,
  ) {
    if (typeof endpoint !== 'string' || endpoint.trim().length === 0) {
      throw new Error('otlp.endpoint must be configured for the OTLP exporter');
    }
    if (!Number.isFinite(timeoutMs) || timeoutMs <= 0) {
      throw new Error('otlp.timeout must be greater than zero');
    }
    if (headers !== undefined && !OtlpTransport.areValidHeaders(headers)) {
      throw new Error('otlp.headers must be a mapping of string header names to string values');
    }

    try {
      this.endpoint = new URL(endpoint);
    } catch {
      throw new Error('otlp.endpoint must be a valid HTTP or HTTPS URL');
    }

    if (this.endpoint.protocol !== 'http:' && this.endpoint.protocol !== 'https:') {
      throw new Error('otlp.endpoint must use the HTTP or HTTPS protocol');
    }

    this.timeoutMs = timeoutMs;
    this.headers = { ...(headers ?? {}) };
    this.onDeliveryFailure = onDeliveryFailure;
    this.retainFailedDeliveries = retainFailedDeliveries;
  }

  /**
   * Export one final DT3 event using the OTLP Logs JSON request format.
   *
   * @param event - Already-masked and validation-processed canonical log event.
   * @throws OtlpTransportError if the request cannot be initiated.
   */
  public export(event: LogEvent): void {
    if (this.closed) {
      throw new OtlpTransportError('OTLP transport is closed', {
        code: Dt3ErrorCode.TransportClosed,
        retryable: false,
      });
    }

    const payload = JSON.stringify(OtlpTransport.toOtlpPayload(event));
    const headers = Object.fromEntries(
      Object.entries(this.headers).filter(([name]) => name.toLowerCase() !== 'content-type'),
    );
    headers['Content-Type'] = 'application/json';

    const requestFactory = this.endpoint.protocol === 'https:' ? requestSecurely : request;
    let resolveDelivery!: () => void;
    let rejectDelivery!: (error: OtlpTransportError) => void;
    const delivery = new Promise<void>((resolve, reject) => {
      resolveDelivery = resolve;
      rejectDelivery = reject;
    });
    this.inFlight.add(delivery);

    // A fail-closed logger needs a settled rejection available to a later
    // flush boundary. Fail-open delivery failures are reported immediately,
    // then released so a long-lived logger cannot retain failed requests.
    void delivery
      .catch((deliveryError: unknown) => {
        if (this.retainFailedDeliveries) {
          this.failedDeliveries.set(delivery, deliveryError);
        }
        try {
          this.onDeliveryFailure?.(deliveryError);
        } catch {
          // The observer must never create an unhandled rejection.
        }
      });
    void delivery.finally(() => this.inFlight.delete(delivery)).catch(() => undefined);

    try {
      const outgoingRequest = requestFactory(
        this.endpoint,
        {
          method: 'POST',
          headers,
          timeout: this.timeoutMs,
        },
        (response) => {
          // Drain the response so the connection is not retained unnecessarily.
          response.resume();

          if (response.statusCode === undefined || response.statusCode < 200 || response.statusCode >= 300) {
            rejectDelivery(
              new OtlpTransportError(`OTLP export failed with status ${response.statusCode ?? 'unknown'}`, {
                code: Dt3ErrorCode.TransportRejected,
                retryable: (response.statusCode ?? 0) >= 500,
              }),
            );
            return;
          }

          resolveDelivery();
        },
      );

      outgoingRequest.once('timeout', () => {
        outgoingRequest.destroy(
          new OtlpTransportError('OTLP export request timed out', {
            code: Dt3ErrorCode.TransportTimeout,
            retryable: true,
          }),
        );
      });
      outgoingRequest.once('error', (error: Error) => {
        // Transport errors intentionally omit the event payload to avoid exposing data.
        rejectDelivery(
          error instanceof OtlpTransportError
            ? error
            : new OtlpTransportError('OTLP export request failed', {
                code: Dt3ErrorCode.TransportUnavailable,
                retryable: true,
                cause: error,
              }),
        );
      });
      outgoingRequest.write(payload, 'utf8');
      outgoingRequest.end();
    } catch {
      rejectDelivery(
        new OtlpTransportError('OTLP export request could not be initiated', {
          code: Dt3ErrorCode.TransportUnavailable,
          retryable: true,
        }),
      );
    }
  }

  /**
   * Wait for OTLP requests that were in flight when this method was called
   * and consume previously settled failures not yet observed by a flush.
   *
   * @returns A promise resolving after all captured requests settle, rejecting
   * with the first sanitized delivery failure.
   */
  public async flush(): Promise<void> {
    if (this.closed) {
      throw new OtlpTransportError('OTLP transport is closed', {
        code: Dt3ErrorCode.TransportClosed,
        retryable: false,
      });
    }

    const pending = [...this.inFlight];
    const retainedFailures = [...this.failedDeliveries.entries()];
    if (pending.length === 0 && retainedFailures.length === 0) {
      return;
    }

    const results = await Promise.allSettled(pending);
    const failure = results.find(
      (result): result is PromiseRejectedResult => result.status === 'rejected',
    );
    const retainedFailure = retainedFailures[0]?.[1];

    // Every request captured at this boundary, including one that rejected
    // while awaiting `allSettled`, has now been observed by this flush.
    for (const [delivery] of retainedFailures) {
      this.failedDeliveries.delete(delivery);
    }
    for (const delivery of pending) {
      this.failedDeliveries.delete(delivery);
    }

    if (retainedFailure !== undefined) {
      throw retainedFailure;
    }
    if (failure) {
      throw failure.reason;
    }
  }

  /**
   * Enter the terminal transport state.
   *
   * The OTLP transport has no persistent resource to release, but closing is
   * idempotent and prevents new exports and flushes.
   */
  public close(): void {
    this.closed = true;
  }

  /**
   * Map a final DT3 event into a standards-shaped OTLP Logs JSON export body.
   *
   * @param event - Final canonical DT3 log event.
   * @returns An OTLP Logs JSON request containing one log record.
   */
  public static toOtlpPayload(event: LogEvent): Record<string, unknown> {
    const eventData = { ...event };
    const severityText = String(eventData.severity ?? 'INFO').toUpperCase();
    const logRecord: Record<string, unknown> = {
      timeUnixNano: String(OtlpTransport.timestampToNanoseconds(eventData.timestamp)),
      severityNumber: OtlpTransport.severityNumbers[severityText] ?? 9,
      severityText,
      body: { stringValue: String(eventData.message ?? '') },
    };

    const logAttributes = OtlpTransport.logAttributes(eventData);
    if (logAttributes.length > 0) {
      logRecord.attributes = logAttributes;
    }

    const scopeAttributes: Record<string, unknown>[] = [];
    if ('sdk.name' in eventData) {
      scopeAttributes.push(OtlpTransport.attribute('dt3.sdk.name', eventData['sdk.name']));
    }
    if ('sdk.version' in eventData) {
      scopeAttributes.push(OtlpTransport.attribute('dt3.sdk.version', eventData['sdk.version']));
    }

    const scopeLog: Record<string, unknown> = { logRecords: [logRecord] };
    if (scopeAttributes.length > 0) {
      scopeLog.scope = { name: 'dt3.logger', attributes: scopeAttributes };
    }

    return {
      resourceLogs: [
        {
          resource: { attributes: OtlpTransport.resourceAttributes(eventData) },
          scopeLogs: [scopeLog],
        },
      ],
    };
  }

  private static resourceAttributes(event: LogEvent): Record<string, unknown>[] {
    const attributes: Record<string, unknown>[] = [];
    for (const key of [
      'service.name',
      'service.version',
      'deployment.environment',
      'tenant.id',
      'tenant.name',
    ]) {
      if (key in event) {
        attributes.push(OtlpTransport.attribute(key, event[key]));
      }
    }
    return attributes;
  }

  private static logAttributes(event: LogEvent): Record<string, unknown>[] {
    const excluded = new Set([
      'timestamp',
      'severity',
      'message',
      'service.name',
      'service.version',
      'deployment.environment',
      'tenant.id',
      'tenant.name',
      'sdk.name',
      'sdk.version',
    ]);

    return Object.entries(event)
      .filter(([key]) => !excluded.has(key))
      .map(([key, value]) => OtlpTransport.attribute(key, value));
  }

  private static attribute(key: string, value: unknown): Record<string, unknown> {
    return { key, value: OtlpTransport.anyValue(value) };
  }

  private static anyValue(value: unknown): Record<string, unknown> {
    if (typeof value === 'boolean') {
      return { boolValue: value };
    }
    if (typeof value === 'number') {
      return Number.isInteger(value) ? { intValue: String(value) } : { doubleValue: value };
    }
    if (typeof value === 'string') {
      return { stringValue: value };
    }
    if (value === null || value === undefined) {
      return { stringValue: 'null' };
    }
    if (Array.isArray(value)) {
      return { arrayValue: { values: value.map((item) => OtlpTransport.anyValue(item)) } };
    }
    if (typeof value === 'object') {
      return {
        kvlistValue: {
          values: Object.entries(value as Record<string, unknown>).map(([key, item]) =>
            OtlpTransport.attribute(key, item),
          ),
        },
      };
    }

    return { stringValue: String(value) };
  }

  private static timestampToNanoseconds(value: unknown): bigint {
    const nowNanoseconds = (): bigint => BigInt(Date.now()) * 1_000_000n;

    if (typeof value !== 'string') {
      return nowNanoseconds();
    }

    const timestamp = Date.parse(value);
    return Number.isNaN(timestamp) ? nowNanoseconds() : BigInt(timestamp) * 1_000_000n;
  }

  private static areValidHeaders(headers: Headers): boolean {
    return Object.entries(headers).every(
      ([name, value]) =>
        typeof name === 'string' &&
        name.trim().length > 0 &&
        !/[\r\n]/.test(name) &&
        typeof value === 'string' &&
        !/[\r\n]/.test(value),
    );
  }
}

File: packages/python/dt3_sdk/otlp_transport.py
"""Synchronous OTLP/HTTP JSON transport for final DT3 structured log events."""

from __future__ import annotations

import json
import threading
from datetime import datetime, timezone
from typing import Any, Mapping, Optional
from urllib.error import HTTPError, URLError
from urllib.request import Request, urlopen

from .errors import Dt3ErrorCode, Dt3TransportError


class OtlpTransportError(Dt3TransportError):
    """Raised when the OTLP transport cannot successfully export an event."""

    code = Dt3ErrorCode.TRANSPORT_UNAVAILABLE
    retryable = True


class OtlpTransport:
    """Synchronously export final DT3 log events through OTLP/HTTP JSON."""

    _SEVERITY_NUMBERS = {
        "TRACE": 1,
        "DEBUG": 5,
        "INFO": 9,
        "WARN": 13,
        "WARNING": 13,
        "ERROR": 17,
        "FATAL": 21,
    }

    def __init__(
        self,
        endpoint: str,
        timeout: float = 10.0,
        headers: Optional[Mapping[str, str]] = None,
    ) -> None:
        """Create an OTLP/HTTP JSON transport.

        Args:
            endpoint: OTLP Logs HTTP endpoint, commonly ending in ``/v1/logs``.
            timeout: Maximum number of seconds allowed for an export request.
            headers: Optional request headers merged with OTLP JSON content type.

        Raises:
            ValueError: If the endpoint, timeout, or configured headers are invalid.
        """
        if not isinstance(endpoint, str) or not endpoint.strip():
            raise ValueError("otlp.endpoint must be configured for the OTLP exporter")
        if timeout <= 0:
            raise ValueError("otlp.timeout must be greater than zero")
        if headers is not None and not isinstance(headers, Mapping):
            raise ValueError(
                "otlp.headers must be a mapping of string header names to string values"
            )

        validated_headers: dict[str, str] = {}
        for name, value in (headers or {}).items():
            if not isinstance(name, str) or not name.strip():
                raise ValueError(
                    f"otlp.headers contains an invalid header name: {name!r}"
                )
            if "\r" in name or "\n" in name:
                raise ValueError("otlp.headers contains an invalid header name")
            if not isinstance(value, str):
                raise ValueError(
                    f"otlp.headers[{name!r}] must have a string header value; got {value!r}"
                )
            if "\r" in value or "\n" in value:
                raise ValueError(f"otlp.headers[{name!r}] contains an invalid header value")
            validated_headers[name] = value

        self._endpoint = endpoint.strip()
        self._timeout = timeout
        self._headers = validated_headers
        self._lock = threading.Lock()
        self._closed = False

    # PUBLIC_INTERFACE
    def export(self, event: Mapping[str, Any]) -> None:
        """Synchronously export one final DT3 event as an OTLP Logs JSON payload.

        Args:
            event: The already-masked and validation-processed canonical log event.

        Raises:
            RuntimeError: If the transport has already been closed.
            OtlpTransportError: If the request fails or receives a non-2xx status.
            TypeError: If the final event cannot be JSON serialized.
        """
        with self._lock:
            if self._closed:
                raise RuntimeError("OTLP transport is closed")

            payload = json.dumps(
                self.to_otlp_payload(event),
                ensure_ascii=False,
                separators=(",", ":"),
            ).encode("utf-8")
            headers = {
                name: value
                for name, value in self._headers.items()
                if name.lower() != "content-type"
            }
            headers["Content-Type"] = "application/json"
            request = Request(
                self._endpoint,
                data=payload,
                headers=headers,
                method="POST",
            )

            try:
                with urlopen(request, timeout=self._timeout) as response:
                    status = response.getcode()
                    if status < 200 or status >= 300:
                        raise OtlpTransportError(
                            f"OTLP export failed with status {status}",
                            code=Dt3ErrorCode.TRANSPORT_REJECTED,
                            retryable=status >= 500,
                        )
            except HTTPError as error:
                raise OtlpTransportError(
                    f"OTLP export failed with status {error.code}",
                    code=Dt3ErrorCode.TRANSPORT_REJECTED,
                    retryable=error.code >= 500,
                ) from error
            except URLError as error:
                reason = str(getattr(error, "reason", "request failed"))
                raise OtlpTransportError(
                    f"OTLP export request failed: {reason}"
                ) from error
            except TimeoutError as error:
                raise OtlpTransportError(
                    "OTLP export request timed out",
                    code=Dt3ErrorCode.TRANSPORT_TIMEOUT,
                    retryable=True,
                ) from error

    # PUBLIC_INTERFACE
    @classmethod
    def to_otlp_payload(cls, event: Mapping[str, Any]) -> dict[str, Any]:
        """Map one final DT3 LogEvent into an OTLP Logs JSON export request.

        Args:
            event: The final canonical DT3 log event.

        Returns:
            A standards-shaped OTLP/HTTP JSON request body containing one log record.
        """
        event_data = dict(event)
        resource_attributes = cls._resource_attributes(event_data)
        log_attributes = cls._log_attributes(event_data)
        timestamp_ns = cls._timestamp_to_nanoseconds(event_data.get("timestamp"))
        severity_text = str(event_data.get("severity", "INFO")).upper()
        severity_number = cls._SEVERITY_NUMBERS.get(severity_text, 9)

        log_record: dict[str, Any] = {
            "timeUnixNano": str(timestamp_ns),
            "severityNumber": severity_number,
            "severityText": severity_text,
            "body": {"stringValue": str(event_data.get("message", ""))},
        }
        if log_attributes:
            log_record["attributes"] = log_attributes

        scope_attributes = []
        if "sdk.name" in event_data:
            scope_attributes.append(
                cls._attribute("dt3.sdk.name", event_data["sdk.name"])
            )
        if "sdk.version" in event_data:
            scope_attributes.append(
                cls._attribute("dt3.sdk.version", event_data["sdk.version"])
            )

        scope_log: dict[str, Any] = {"logRecords": [log_record]}
        if scope_attributes:
            scope_log["scope"] = {"name": "dt3.logger", "attributes": scope_attributes}

        return {
            "resourceLogs": [
                {
                    "resource": {"attributes": resource_attributes},
                    "scopeLogs": [scope_log],
                }
            ]
        }

    # PUBLIC_INTERFACE
    def flush(self) -> None:
        """Provide the synchronous OTLP lifecycle flush operation.

        Exports complete before ``export`` returns, so no buffered output remains.
        """
        with self._lock:
            if self._closed:
                raise RuntimeError("OTLP transport is closed")

    # PUBLIC_INTERFACE
    def close(self) -> None:
        """Close this transport and prevent future exports.

        The operation is idempotent because OTLP/HTTP exports do not retain
        persistent connections or background resources.
        """
        with self._lock:
            self._closed = True

    @classmethod
    def _resource_attributes(cls, event: Mapping[str, Any]) -> list[dict[str, Any]]:
        """Build OTLP resource attributes from DT3 service metadata."""
        attributes: list[dict[str, Any]] = []
        for key in (
            "service.name",
            "service.version",
            "deployment.environment",
            "tenant.id",
            "tenant.name",
        ):
            if key in event:
                attributes.append(cls._attribute(key, event[key]))
        return attributes

    @classmethod
    def _log_attributes(cls, event: Mapping[str, Any]) -> list[dict[str, Any]]:
        """Build OTLP log-record attributes from non-resource DT3 fields."""
        excluded = {
            "timestamp",
            "severity",
            "message",
            "service.name",
            "service.version",
            "deployment.environment",
            "tenant.id",
            "tenant.name",
            "sdk.name",
            "sdk.version",
        }
        return [
            cls._attribute(key, value)
            for key, value in event.items()
            if key not in excluded
        ]

    @staticmethod
    def _attribute(key: str, value: Any) -> dict[str, Any]:
        """Convert a DT3 value into an OTLP JSON key-value attribute."""
        return {"key": str(key), "value": OtlpTransport._any_value(value)}

    @staticmethod
    def _any_value(value: Any) -> dict[str, Any]:
        """Encode a Python value using an OTLP JSON AnyValue shape."""
        if isinstance(value, bool):
            return {"boolValue": value}
        if isinstance(value, int):
            return {"intValue": str(value)}
        if isinstance(value, float):
            return {"doubleValue": value}
        if isinstance(value, str):
            return {"stringValue": value}
        if value is None:
            return {"stringValue": "null"}
        if isinstance(value, Mapping):
            return {
                "kvlistValue": {
                    "values": [
                        OtlpTransport._attribute(str(key), item)
                        for key, item in value.items()
                    ]
                }
            }
        if isinstance(value, (list, tuple)):
            return {
                "arrayValue": {
                    "values": [OtlpTransport._any_value(item) for item in value]
                }
            }
        return {"stringValue": str(value)}

    @staticmethod
    def _timestamp_to_nanoseconds(value: Any) -> int:
        """Convert an RFC 3339 timestamp to OTLP Unix epoch nanoseconds."""
        if not isinstance(value, str):
            return int(datetime.now(timezone.utc).timestamp() * 1_000_000_000)

        try:
            normalized = value.replace("Z", "+00:00")
            timestamp = datetime.fromisoformat(normalized)
            if timestamp.tzinfo is None:
                timestamp = timestamp.replace(tzinfo=timezone.utc)
            return int(timestamp.timestamp() * 1_000_000_000)
        except ValueError:
            return int(datetime.now(timezone.utc).timestamp() * 1_000_000_000)
